src/Blocks/BlockSapling.h

```diff
@@ -34,17 +35,144 @@ class cBlockSaplingHandler :
 	virtual void OnUpdate(cChunkInterface & cChunkInterface, cWorldInterface & a_WorldInterface, cBlockPluginInterface & a_PluginInterface, cChunk & a_Chunk, int a_RelX, int a_RelY, int a_RelZ) override
 	{
 		NIBBLETYPE Meta = a_Chunk.GetMeta(a_RelX, a_RelY, a_RelZ);
-		
-		if ((Meta & 0x08) != 0)
+		NIBBLETYPE Light = std::max(a_Chunk.GetBlockLight(a_RelX, a_RelY, a_RelZ), a_Chunk.GetTimeAlteredLight(a_Chunk.GetSkyLight(a_RelX, a_RelY, a_RelZ)));
+
+		// Only grow if we have the right amount of light
+		if (Light > 8)
+		{
+			cFastRandom random;
+			// Only grow if we are in the right growth stage and have the right amount of space around them.
+			if (((Meta & 0x08) != 0) && (random.NextInt(99) < 45) && CanGrowAt(a_Chunk, a_RelX, a_RelY, a_RelZ, Meta))
+			{
+				int BlockX = a_RelX + a_Chunk.GetPosX() * cChunkDef::Width;
+				int BlockZ = a_RelZ + a_Chunk.GetPosZ() * cChunkDef::Width;
+				a_Chunk.GetWorld()->GrowTree(BlockX, a_RelY, BlockZ);
+			}
+			// Only move to the next growth stage if we haven't gone there yet
+			else if (((Meta & 0x08) == 0) && (random.NextInt(99) < 45))
+			{
+				a_Chunk.SetMeta(a_RelX, a_RelY, a_RelZ, Meta | 0x08);
+			}
+		}
+	}
+
+	bool CanGrowAt(cChunk & a_Chunk, int a_RelX, int a_RelY, int a_RelZ, NIBBLETYPE a_Meta)
+	{
+		a_Meta = a_Meta & 0x07;
+		int CheckHeight = 0;
+		bool LargeTree = false;
+
+		// Get the height to check against
+		switch (a_Meta)
 		{
-			int BlockX = a_RelX + a_Chunk.GetPosX() * cChunkDef::Width;
-			int BlockZ = a_RelZ + a_Chunk.GetPosZ() * cChunkDef::Width;
-			a_Chunk.GetWorld()->GrowTree(BlockX, a_RelY, BlockZ);
+			case E_META_SAPLING_APPLE:
+			{
+				CheckHeight = 5;
+				break;
+			}
+			case E_META_SAPLING_CONIFER:
+			{
+				CheckHeight = 7;
+				if (IsLargeTree(a_Chunk, a_RelX, a_RelY, a_RelZ, a_Meta))
+				{
+					CheckHeight = 16;
+					LargeTree = true;
+				}
+				break;
+			}
+			case E_META_SAPLING_BIRCH:
+			{
+				CheckHeight = 6;
+				break;
+			}
+			case E_META_SAPLING_JUNGLE:
+			{
+				CheckHeight = 7;
+				if (IsLargeTree(a_Chunk, a_RelX, a_RelY, a_RelZ, a_Meta))
+				{
+					CheckHeight = 13;
```

Comment:
Is a large tree only 13 blocks high? I thought it was much higher. Lemme check :)